### DISCUSSION TITLE
When overloading, what priority is the overload called?

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#6924] [Q&A] [answered]
I have a class to help me customize the format to  numbers
```csharp	
public struct Float : IFormattable
{
	float value;
	public override string ToString() => ToString(null, null);
	public string ToString(string format, IFormatProvider formatProvider)...

	public static implicit operator Float(float f)
	{
		return new Float { value = f };
	}
	public static implicit operator float(Float f)
	{
		return f.value;
	}
}
```
Why doesn't he think that my parameter is of type object, but he implicitly converts it to float, and then uses float overloading?
```csharp
Float f = 12345678;
Console.WriteLine(f.ToString());		//12.3M
Console.WriteLine(f);				//123456789
Console.WriteLine($"{f}");			//12.3M
```

**Comment by CyrusNajmabadi** [comment, marked as answer]
Overload resolution is specified here: https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1164-overload-resolution

**Reply by CyrusNajmabadi** [reply]
The parts of relevance are:

```
Given an argument list A with a set of argument expressions {E₁, E₂, ..., Eᵥ} and two applicable function members Mᵥ and Mₓ 
with parameter types {P₁, P₂, ..., Pᵥ} and {Q₁, Q₂, ..., Qᵥ}, Mᵥ is defined to be a better function member than Mₓ if

for each argument, the implicit conversion from Eᵥ to Qᵥ is not better than the implicit conversion from Eᵥ to Pᵥ, and
for at least one argument, the conversion from Eᵥ to Pᵥ is better than the conversion from Eᵥ to Qᵥ.
```

So the question is whether the conversion of `Float` to `float` is better than the conversion from `Float` to `object`.

THis is determined here: https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11646-better-conversion-target

```
Given two types T₁ and T₂, T₁ is a better conversion target than T₂ if one of the following holds:
An implicit conversion from T₁ to T₂ exists and no implicit conversion from T₂ to T₁ exists 
```

So here we have `float` and `object` as conversion targets.  `float` is considered better as there is an implicit conversion from `float` to `object` but no implicit conversion from `object` to `float`.

Put another way, `float` seems like a more precise/narrow/specific match, and `object` is a less precise/wider/less-specific match.  We prefer the more precise one, and so the `float` overload is preferred.
